### PR TITLE
feat: 15146 Added `calculateHash`, `setHash` and `getHash` methods to `State` interface

### LINKED ISSUE #15146: Create a formal API method in the State API for hashing
Add a method, `Hash hash()` to the State API that hashes that instance of the state. If the state is mutable, throw an exception. If the state is already hashed, just return the existing hash.

### PR DESCRIPTION
**Pull request by imalygin**
**Description**:

This PR adds hash-related methods to `State` interface and provides an implementation of these methods in `MerkleStateRoot`. In fact, only `calculateHash` implementation is required, as the other methods are already implemented  by a parent class (`AbstractMerkleNode`). 

**Related issue(s)**:

Fixes #15146 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java [hunk 1]
```diff
@@ -72,8 +75,34 @@ default void unregisterCommitListener(@NonNull final StateChangeListener listene
...
+    @Nullable
+    @Override
+    default Hash getHash() {
+        throw new UnsupportedOperationException();
+    }
+
+    /***
+     * {$inheritDoc}
+     */
+    default void setHash(@NonNull Hash hash) {
```
Comment (anthony-swirldslabs) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java:98]: Nit: `@Override` is missing.

**Review by anthony-swirldslabs — DISMISSED** [review]

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java (hunk 1 above)
Comment (imalygin) [reply platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java]: Fixed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 2]
```diff
@@ -1042,4 +1045,24 @@ private WritablePlatformStateStore writablePlatformStateStore() {
...
+     */
+    @Override
+    public void calculateHash() {
+        throwIfMutable("Hashing should only be done on immutable states");
+        throwIfDestroyed("Hashing should not be done on destroyed states");
+        if (getHash() != null) {
+            return;
+        }
+        try {
+            MerkleCryptoFactory.getInstance().digestTreeAsync(this).get();
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:1060]: This class is deprecated, is it possible to pass in `MerkleCryptography`?

**Review by litt3 — APPROVED** [review]

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java [hunk 3]
```diff
@@ -72,8 +75,35 @@ default void unregisterCommitListener(@NonNull final StateChangeListener listene
         throw new UnsupportedOperationException();
     }
 
+    /***
+     * {$inheritDoc}
```
Comment (mhess-swl) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java:79]: Nit: shouldn't this be `@`? 

```suggestion
     * {@inheritDoc}
```

**Review by mhess-swl — APPROVED** [review]
`inheritDoc` comment can be fixed in a future PR if needed. The release branch is scheduled to be created in under 2 hours

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java [hunk 4]
```diff
@@ -1042,4 +1045,24 @@ private WritablePlatformStateStore writablePlatformStateStore() {
...
+        } catch (final ExecutionException e) {
+            logger.fatal(EXCEPTION.getMarker(), "Exception occurred during hashing", e);
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:1062]: If we log a fatal exception, we should do more (e.g. throw a runtime exception, or error, or take some other action to either recover or stop the system).

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java [hunk 5]
```diff
@@ -1067,4 +1069,68 @@ void migrate_platform_absent() {
...
+    @DisplayName("Hashing test")
+    class HashingTest {
+
+        @BeforeEach
+        void setUp() {
+            setupAnimalMerkleMap();
+            setupSingletonCountry();
+            setupSteamQueue();
+
+            add(fruitMerkleMap, fruitMetadata, A_KEY, APPLE);
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java:1083]: Do we need to call `setupFruitMerkleMap` before adding to it?

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java [hunk 6]
```diff
@@ -72,8 +75,35 @@ default void unregisterCommitListener(@NonNull final StateChangeListener listene
...
+    default void setHash(@NonNull Hash hash) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Hashes the state on demand if it is not already hashed. If the state is already hashed, this method is a no-op.
+     */
+    default void calculateHash() {
+        throw new UnsupportedOperationException();
+    }
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java:108]: Why have default methods if they just throw?
Would it not be better to have the compiler require implementing classes to provide these methods?

**Review by jsync-swirlds — APPROVED** [review]
A few items that can be addressed in future PRs.

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 4 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java:1062]: It's not that dramatic in this case, so I'm going to change `fatal` to `error`

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java:1083]: Yes, I'll added it

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java (hunk 6 above)
Comment (imalygin) [reply platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java:108]: They have this implementation to ease the burden on the test and helper classes that happened to implement `State` interface but don't need that functionality at all.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java (hunk 2 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/MerkleStateRoot.java]: Done

File: platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java (hunk 3 above)
Comment (imalygin) [reply platform-sdk/swirlds-state-api/src/main/java/com/swirlds/state/State.java]: yes, fixed

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java (hunk 5 above)
Comment (imalygin) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/MerkleStateRootTest.java:1091]: Fixed

**Review by jsync-swirlds — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by mhess-swl — APPROVED** [review]

**Review by anthony-swirldslabs — APPROVED** [review]

**Review by lpetrovic05 — APPROVED** [review]

**Review by litt3 — APPROVED** [review]

**Review by tinker-michaelj — APPROVED** [review]
Makes sense to me, ty @imalygin !

**Pull request merged**